test: cover addrv2 anchors by adding TorV3 to CAddress in messages.py

pinheadmz opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/27140

Adds test coverage for https://github.com/bitcoin/bitcoin/pull/20516 to ensure that https://github.com/bitcoin/bitcoin/issues/20511 is completed and may be closed.

This PR adds a test case to `feature_anchors.py` where an onion v3 address is set as a blocks-only relay peer and then shutdown, ensuring that the address is saved to anchors.dat in addrv2 format. We then ensure that bitcoin attempts to reconnect to that anchor address on restart.

To compute the addrv2 serialization of the onion v3 address, I added logic to `CAddress` in `messages.py`. This new logic is covered by extending `p2p_addrv2_relay.py` to include an onion v3 address. Future work will be adding coverage for ipv6, torv2 and cjdns in these modules and also `feature_proxy.py`

Also includes de/serialization unit test for `CAddress` in test framework.

brunoerg committed:
test: add support for all networks in CAddress in messages.py

Also removes TorV2 from messages.py
See https://github.com/bitcoin/bitcoin/pull/22050

Co-authored-by: Matthew Zipkin <pinheadmz@gmail.com>

brunoerg commented:
Did you take a look at #27295?

brunoerg commented on test/functional/feature_anchors.py:
```diff
@@ -78,6 +81,60 @@ def run_test(self):
+
+        self.log.info("Check peer info")
+        peer_info = self.nodes[0].getpeerinfo()
+        assert len(peer_info) == 1
```
I think you could use `assert_equal` here and in the other similar cases

```suggestion
        assert_equal(len(peer_info), 1)
```

pinheadmz replied:
sure, done thanks!

brunoerg reviewed (commented):
Concept ACK

earuak commented:
It is important to include test cases to ensure that the code is working correctly and that adding new features like logic to compute the addrv2 serialization of the onion v3 address can help improve the functionality of Bitcoin. Also, it's always good to keep working on future coverage for other modules and features.

willcl-ark commented on test/functional/feature_anchors.py:
```diff
@@ -78,6 +81,60 @@ def run_test(self):
+            # even though we never connected to it
+            new_data = bytearray(data)[:-32]
+            new_data[services_index] = P2P_SERVICES
+            new_data_hash = hashlib.sha256(hashlib.sha256(new_data).digest()).digest()
```
Import helper from _test_framework.messages_?

```suggestion
            new_data_hash = hash256(new_data)

```

willcl-ark commented on test/functional/feature_anchors.py:
```diff
@@ -4,11 +4,14 @@
-from test_framework.p2p import P2PInterface
+from test_framework.p2p import P2PInterface, P2P_SERVICES
+from test_framework.socks5 import Socks5Configuration, Socks5Server
+from test_framework.messages import CAddress
```
```suggestion
from test_framework.messages import CAddress, hash256
```

willcl-ark commented on test/functional/feature_anchors.py:
```diff
@@ -78,6 +81,60 @@ def run_test(self):
+        onion_conf.auth = True
+        onion_conf.unauth = True
+        onion_conf.addr = ('127.0.0.1', 10000)
+        onion_conf.keep_alive = True
```
Why do we need to keep the `Socks5Server` alive to ensure success when performing `"Check for addrv2 addresses in anchors.dat"`?

I changed this to `False` and saw intermittent failures on Line 123

```
assert_equal(data[services_index], 0x00) # services == NONE
```

pinheadmz replied:
Good question. The Socks5 server we use in the test framework by default disconnects immediately after the handshake and sometimes would not register as a connected peer:

https://cirrus-ci.com/task/6075793026056192?logs=ci#L3820-L3830

pinheadmz commented:
Thanks for the review Will, nits addressed and pushed

vasild commented:
Concept ACK, I guess we can forget about Torv2 and completely drop support about it from everywhere as if it never existed. The Tor network does not support those anymore.

vasild commented on test/functional/feature_anchors.py:
```diff
@@ -78,6 +80,60 @@ def run_test(self):
+        onion_conf = Socks5Configuration()
+        onion_conf.auth = True
+        onion_conf.unauth = True
+        onion_conf.addr = ('127.0.0.1', 10000)
```
This could collide with another port randomly assigned by the testing framework (even if it looks like that it currently cannot, it may do in the future) and also could collide with another instance of this test run in parallel. Better use `p2p_port()`.

pinheadmz replied:
thanks, fixed.

vasild commented on test/functional/feature_anchors.py:
```diff
@@ -78,6 +80,60 @@ def run_test(self):
+            anchors2 = data.hex()
+            assert expected_pubkey in anchors2
+
+        with open(node_anchors_path, "wb") as file_handler:
+            # Cheat: modify service flags for this address
+            # even though we never connected to it
+            new_data = bytearray(data)[:-32]
+            new_data[services_index] = P2P_SERVICES
+            new_data_hash = hash256(new_data)
+            file_handler.write(new_data + new_data_hash)
```
Why is it necessary to modify the services?

pinheadmz committed:
test: add keep_alive option to socks5 proxy in test_framework

The Socks5 server we use in the test framework would disconnect
by default immediately after the handshake and sometimes would
not register as a connected peer by bitcoind.

pinheadmz committed:
test: cover TorV3 address in p2p_addrv2_relay

pinheadmz committed:
test: cover addrv2 support in anchors.dat with a TorV3 address

pinheadmz commented:
@brunoerg @jonatack thanks for your reviews, pushing to 455f8c4f0873df25eadd0cd59e67d5eb2a568621:

Includes and expands on https://github.com/bitcoin/bitcoin/pull/27295 including a unit test to de- and re-serialize CAddress from every network type. TorV2 is also completely dropped from the test framework.

vasild reviewed (commented):
Approach ACK b5e0398d8ced2ee998756e67fb90bc8d604d4ec3

pinheadmz commented on test/functional/test_framework/messages.py:
```diff
@@ -285,33 +299,56 @@ def deserialize_v2(self, f):
 
         self.net = struct.unpack("B", f.read(1))[0]
-        assert self.net in (self.NET_IPV4, self.NET_I2P)
+        assert self.net in (self.NET_IPV4, self.NET_IPV6,
+                            self.NET_I2P, self.NET_CJDNS,
+                            self.NET_TORV3)
```
yes thanks

pinheadmz commented on test/functional/test_framework/socks5.py:
```diff
@@ -121,7 +122,8 @@ def handle(self):
             logger.exception("socks5 request handling failed.")
             self.serv.queue.put(e)
         finally:
-            self.conn.close()
+            if not self.serv.keep_alive:
+                self.conn.close()
```
Good question, I'm just letting it drop out of scope at the end of the test, but [looking into it more](https://docs.python.org/3/library/threading.html#threading.Thread.daemon) it may actually be staying alive until the entire test runner exits.

Should I add a manual close function?

pinheadmz replied:
update: no longer needed

pinheadmz replied:
update update: manually calling `onion_proxy.close()` in the test now with a comment

pinheadmz commented on test/functional/p2p_addrv2_relay.py:
```diff
@@ -20,19 +20,29 @@
         addr.ip = f"123.123.123.{i % 256}"
-    addr.port = 8333 + i
+        msg_size += addr.ADDRV2_ADDRESS_LENGTH[addr.NET_IPV4]
+    msg_size += 4 + 1 + 1 + 2 + 1  # time, services, networkID, port, address length byte
```
fantastic, thanks for the patch

pinheadmz commented on test/functional/feature_anchors.py:
```diff
@@ -86,6 +88,60 @@ def run_test(self):
+        onion_conf.keep_alive = True
+        onion_proxy = Socks5Server(onion_conf)
+        onion_proxy.start()
+        self.restart_node(0, extra_args=[f"-onion={onion_conf.addr[0]}:{onion_conf.addr[1]}"])
+
+        self.log.info(f"Add 256-bit-address block-relay-only connections to node")
+        ONION_ADDR = "pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion:8333"
+        self.nodes[0].addconnection(ONION_ADDR, 'block-relay-only')
+
+        self.log.info("Check peer info")
+        peer_info = self.nodes[0].getpeerinfo()
+        assert_equal(len(peer_info), 1)
+        assert_equal(peer_info[0]["network"], "onion")
+        assert_equal(peer_info[0]["connection_type"], "block-relay-only")
```
oh sweet! awesome, i'll remove the socks5.py commit and apply this instead

pinheadmz replied:
https://cirrus-ci.com/task/5531648075235328?logs=ci#L4421

Actually there is still a race condition there. I think the keep_alive is necessary for bitcoind to even acknowledge the blocks only peer exists. We can use debug log or getpeerinfo but seems like either way the socks proxy is closing too soon

pinheadmz commented on test/functional/feature_anchors.py:
```diff
@@ -84,6 +86,54 @@ def run_test(self):
+        self.nodes[0].addconnection(ONION_ADDR, 'block-relay-only')
+
+        self.log.info("Stop node 0")
+        with self.nodes[0].assert_debug_log([f"DumpAnchors: Flush 1 outbound block-relay-only peer addresses to anchors.dat"]):
```
Indeed :-( see https://github.com/bitcoin/bitcoin/pull/27452#discussion_r1253478816 unless @vasild disagrees I'll revert part of my last update and put `keep_alive` back in socks5.py

pinheadmz commented:
@jonatack @vasild thanks for reviewing guys. I put back the `keep_alive` option to ensure the onion peer is logged, expecting successful CI on this push 🤞

jonatack reviewed (commented):
Tested Approach ACK

pinheadmz commented:
@jonatack thank you sir for the great review and suggested patches, all nits addressed

jonatack reviewed (commented):
ACK a5c9e059b7f463f668fa404dfc652791c481fd95

brunoerg reviewed (approved):
crACK ba8ab4fc545800c4fb283a5ff0b19138a0451aba

code lgtm!

willcl-ark reviewed (approved):
ACK ba8ab4fc54

Nice to have this additional coverage for TorV3 addresses 👍🏼 

It's not intruduced by this PR, but I notice a mypy error in _feature_anchors.py_ where a tuple is assigned to `onion_conf.addr` which is initialized to `None` in _socks5.py_.

![image](https://github.com/bitcoin/bitcoin/assets/6606587/c44f86b1-15ab-4d86-9e89-5617866b8dad)

If you do end up re-touching this for some reason, you can fix the warning by specifying the type of `.addr` in _socks5.py_ (which will also fix the other 3 ocurrences of it in the tests):

```diff
diff --git a/test/functional/test_framework/socks5.py b/test/functional/test_framework/socks5.py
index 0ca06a7396..5315f39142 100644
--- a/test/functional/test_framework/socks5.py
+++ b/test/functional/test_framework/socks5.py
@@ -8,6 +8,7 @@ import socket
 import threading
 import queue
 import logging
+from typing import Tuple
 
 logger = logging.getLogger("TestFramework.socks5")
 
@@ -36,7 +37,7 @@ def recvall(s, n):
 class Socks5Configuration():
     """Proxy configuration."""
     def __init__(self):
-        self.addr = None # Bind address (must be set)
+        self.addr: Tuple[str, int] # Bind address (must be set)
         self.af = socket.AF_INET # Bind address family
         self.unauth = False  # Support unauthenticated
         self.auth = False  # Support authentication
```
